library/general/containerresourceratios/src.rego

```diff
@@ -0,0 +1,228 @@
+  not canonify_cpu(cpu_orig)
+  msg := sprintf("container <%v> cpu request <%v> could not be parsed", [container.name, cpu_orig])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  mem_orig := container.resources.requests.memory
+  not canonify_mem(mem_orig)
+  msg := sprintf("container <%v> memory request <%v> could not be parsed", [container.name, mem_orig])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  not container.resources
+  msg := sprintf("container <%v> has no resource limits", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  not container.resources.limits
+  msg := sprintf("container <%v> has no resource limits", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  missing(container.resources.limits, "cpu")
+  msg := sprintf("container <%v> has no cpu limit", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  missing(container.resources.limits, "memory")
+  msg := sprintf("container <%v> has no memory limit", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  not container.resources.requests
+  msg := sprintf("container <%v> has no resource requests", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  missing(container.resources.requests, "cpu")
+  msg := sprintf("container <%v> has no cpu request", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  missing(container.resources.requests, "memory")
+  msg := sprintf("container <%v> has no memory request", [container.name])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  cpu_limits_orig := container.resources.limits.cpu
+  cpu_limits := canonify_cpu(cpu_limits_orig)
+  cpu_requests_orig := container.resources.requests.cpu
+  cpu_requests := canonify_cpu(cpu_requests_orig)
+  cpu_ratio := input.parameters.ratio
+  cpu_calc_ratio := cpu_limits / cpu_requests
+  to_number(cpu_ratio) < to_number(cpu_calc_ratio)
+  msg := sprintf("container <%v> cpu limit <%v> is higher than the maximum allowed ratio of <%v>", [container.name, cpu_limits_orig, cpu_ratio])
+}
+
+general_violation[{"msg": msg, "field": field}] {
+  container := input.review.object.spec[field][_]
+  mem_limits_orig := container.resources.limits.memory
+  mem_requests_orig := container.resources.requests.memory
+  mem_limits := canonify_mem(mem_limits_orig)
+  mem_requests := canonify_mem(mem_requests_orig)
+  mem_ratio := input.parameters.ratio
+  mem_calc_ratio := mem_limits / mem_requests
```

Comment:
It seems to me that requiring an exact match after applying a specific ratio could be vulnerable to rounding errors.

Could this instead be an integer multiple? i.e. mem_limits = 2 * mem_requests? If all values are required to be integers we eliminate the possibility of rounding errors.

Another option may be to build in some error tolerance.

Reply:
I'm not sure I understand this comment.
This is not a boolean condition, it's an assignment.
The condition only happens in the next line, and is not exact (rather it's a lt operator)

I've altered the code to use multiplication sense the underlying result remains the same, but, just out of curiosity,  I would be happy if you can give me an example of an edge case that might create such an error.

Reply:
Not sure why I thought there was an `==` somewhere. I'm guessing I got the impression that the constraint was `==` and saw what I expected to see?

 It's a bit fiddly to get these errors as floating point precision is pretty large, and ratios of integers may not include these errors depending on how they calculate the quotient, but this shows an example of an expected result breaking with addition:

https://play.golang.org/p/KZ3Hw9Z3Rzj

General rule of thumb when creating numbers that need to be compared is that if you can avoid using floating points, it's best to do so.

Reply:
This Python tutorial goes into a bit more detail:

https://docs.python.org/3/tutorial/floatingpoint.html

Reply:
Interesting that the same does not break in Rego (which is backed by golang):

https://play.openpolicyagent.org/p/EU1hOWjW1i

I wonder if OPA is using an exact floating-point library under the hood. I asked @tsandall

Reply:
Thanks! Appreciate the detailed explanation, I'll go over these examples soon.